### PR TITLE
Enables --encode_unicode_case option for case-aware sentence piece

### PR DESCRIPTION
**Pull request by emjotde**
* Enables --encode_unicode_case option for case-aware sentence piece
* Example: `This IS a TEST OF THE CASING` gets converted internally to `Tthis Uis a Atest of the casing` before segmentation.
* This is fully reversible.

### CODE REVIEW

File: src/case_encoder.cc [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  } else if(decodeCase) {
+    return std::unique_ptr<CaseEncoder>(new UpperCaseDecoder());
+  } else {
+    return nullptr;
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////
+// Implements finite state automaton for regex: Uu+(sss|p|$)+Uu+(sss|p|$)+(Uu+(sss|p|$)+)+
+// std::regex seems to be causing weird issues with stack overflow etc. on Windows.
```
Comment (snukky) [inline src/case_encoder.cc:35]: Could this be similar to PR https://github.com/marian-nmt/sentencepiece/pull/7 and be solved by defining some variables for regexes? Anyhow, FSA is probably faster, so just making you aware of Rohit's solution.

File: src/builder.cc [hunk 2]
```diff
@@ -37,6 +37,7 @@
 
 #include "normalization_rule.h"
 #include "normalizer.h"
+#include "case_encoder.h"
```
Comment (snukky) [inline src/builder.cc:40]: Nit: seems that these includes were sorted alphabethically.

File: src/case_encoder.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 Google Inc.
```
Comment (snukky) [inline src/case_encoder.cc:1]: Seems weird adding an old copyright header for a new code, but I don't have the solution.

**Review by snukky — APPROVED** [review]
I'm not familiar with SentencePiece code, so only a few generic comments from me.

Would be great to have some documentation on this in https://github.com/marian-nmt/sentencepiece/tree/master/doc - even a very brief description or just a couple of command-line/input/output examples will be useful. This can be added later in a separate PR.

Do I remember correctly that we wanted to add some regression tests for this new feature into Marian regression tests? Basic SentencePiece features, including normalization, etc., are already covered by tests creating different types of vocabs via Marian.

**Review by rjai — APPROVED** [review]

File: src/case_encoder.cc (hunk 1 above)
Comment (rjai) [reply src/case_encoder.cc:35]: This problem had repeated itself and we were getting system Stack Overflows, hence the need to fix it this way.

File: src/case_encoder.cc (hunk 3 above)
Comment (emjotde) [reply src/case_encoder.cc:1]: Meh :) We will figure that out if we ever share back into the original repository.

**Pull request merged**